File: packages/fruitbox-pwa/src/stats/stats-store.ts
export interface GameInfo {
  gamemode: string;
  grid_type: string;
  self_score: number;
  seed: number;
  time_elapsed: number;
  opp_score?: number | null;
}

export interface StatsSummary {
  total_games: number;
  total_time: number;
  vs_wins: number;
  vs_losses: number;
  vs_ties: number;
  random_best: number | null;
  random_best_seed: number | null;
  random_best_time: number | null;
  solvable_best: number | null;
  solvable_best_seed: number | null;
  solvable_best_time: number | null;
}

export interface HistoryRow {
  game_id: string;
  gamemode: string;
  grid_type: string;
  self_score: number;
  opp_score: number | null;
  time_elapsed: number;
  seed: number;
}

export interface StatsStore {
  record(info: GameInfo): string;
  getSummary(): StatsSummary;
  getVsStats(): { wins: number; losses: number; ties: number };
  getHistory(): HistoryRow[];
  exportDatabase(): Uint8Array;
  importDatabase(data: Uint8Array): void;
  persist(): Promise<void>;
}

File: packages/fruitbox-pygame/src/fruitbox_pygame/stats_store.py
from typing import Protocol

from fruitbox_core.models import GameInfo


class StatsStore(Protocol):
    def record(self, info: GameInfo) -> str: ...

    def get_summary(self) -> dict: ...

    def get_vs_stats(self) -> dict: ...

    def get_history(self) -> list[dict]: ...


_EMPTY_SUMMARY = {
    "total_games": 0,
    "total_time": 0,
    "vs_wins": 0,
    "vs_losses": 0,
    "vs_ties": 0,
    "random_best": None,
    "random_best_seed": None,
    "random_best_time": None,
    "solvable_best": None,
    "solvable_best_seed": None,
    "solvable_best_time": None,
}


class NoOpStatsStore:
    """Stats sink for modes or platforms without persistence."""

    def record(self, info: GameInfo) -> str:
        return ""

    def get_summary(self) -> dict:
        return dict(_EMPTY_SUMMARY)

    def get_vs_stats(self) -> dict:
        return {"wins": 0, "losses": 0, "ties": 0}

    def get_history(self) -> list[dict]:
        return []
